core/src/main/scala/org/http4s/util/UrlCoding.scala

```diff
@@ -44,13 +48,42 @@ private[util] trait UrlCodingUtils {
     })
   }
 
+  /**
+    * Percent-encodes a string.  Depending on the parameters, this method is
+    * appropriate for URI or URL form encoding.  Any resulting percent-encodings
+    * are normalized to uppercase.
+    * 
+    * Characeters in `toSkip` are rendered as is.  If a skipped '%' is followed
+    * by exactly two hex digits, they are uppercased for normalization purposes.
+    * 
+    * Space is rendered as `" "` if skipped.  If space is not skipped, it is
+    * encoded as `"+"` when `spaceIsPlus` is true, or percent encoded otherwise.
+    * 
+    * All characters not in toSkip are rendered in a percent encoding, according
+    * to `charset`.
+    */
   def urlEncode(toEncode: String, charset: Charset = Utf8, spaceIsPlus: Boolean = false, toSkip: BitSet = toSkip) = {
-    val in = charset.encode(ensureUppercasedEncodings(toEncode))
+    val in = charset.encode(toEncode)
     val out = CharBuffer.allocate((in.remaining() * 3).ceil.toInt)
     while (in.hasRemaining) {
       val b = in.get() & 0xFF
       if (toSkip.contains(b)) {
         out.put(b.toInt.toChar)
+        if (b == '%' && in.hasRemaining) {
```

Comment:
Only if `'%'` is in the `toSkip` set.  I struggle to think of when this would ever be a good idea.  I think all unreserved characters should always be in `toSkip`, some reserved characters in certain circumstances, and nothing else.

This method has always normalized, and I think it's the right thing to do as long as we allow skipping `'%'`.  It just normalizes too early in 0.14.6.